New module: Add module to add/remove a network static route on windows (windows/win_route)

dlazz opened the pull request:
##### SUMMARY
Adds win_route windows module
Allow to add or remove a network static route to a windows machine.
It uses route ADD and route DELETE command instead of powershell to avoid user interaction.
Ip address and subnet mask verification.
 
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
win_route

##### ANSIBLE VERSION

```
2.2.2.0
```

Commit:
adds win_route windows module

Commit:
fix documentation indent

Commit:
fix documentation pep8 issues

Commit:
fix documentation issues

dagwieers commented on lib/ansible/modules/windows/win_route.ps1:
```diff
@@ -0,0 +1,171 @@
+
+# win_route (Add or remove a network static route)
+
+$params = Parse-Args $args;
```
Please add check-mode support.

dlazz replied:
Is there any place where all this best practices are documented? I found nothing, but maybe I didn't search enough...

dagwieers replied:
Unfortunately there is no such document for Windows modules yet. However we have updated most of the generic Windows modules to follow the most important rules. So if someone starts from the current modules, they should already be fine for most parts !

We do have the following documents on module development:
- http://docs.ansible.com/ansible/dev_guide/developing_modules_best_practices.html
- http://docs.ansible.com/ansible/dev_guide/developing_modules_checklist.html

You will see that some of my recommendations are documented in that second link.

dlazz replied:
Sorry, I'm a newbie in developing ansible module and in contributing in an open source project! Thanks for your suggestion, I'll keep in mind for the next time :)

dagwieers commented on lib/ansible/modules/windows/win_route.ps1:
```diff
@@ -0,0 +1,171 @@
+
+$params = Parse-Args $args;
+
+$destip = Get-Attr $params "destination_ip" -failifempty $true;
```
We now use Get-AnsibleParam and -type for each parameter. (See for example win_file or win_stat)
Trailing semi-colon is not needed in modules.

dagwieers commented on lib/ansible/modules/windows/win_route.ps1:
```diff
@@ -0,0 +1,171 @@
+$gateway = Get-Attr $params "gateway" -failifempty $true;
+$metric = Get-Attr $params "metric" -default 1;
+$state = Get-Attr $params "state" -default "present" -validateSet "present","absent";
+$result = New-Object PSObject @{"changed" = $false; "output" = ""};
```
Nowadays $result is a simple hashtable, not a PSObject.
```powershell
$result = @{
    changed = $false
    output = ''
}
```

dagwieers commented on lib/ansible/modules/windows/win_route.py:
```diff
@@ -0,0 +1,84 @@
+
+EXAMPLES = r'''
+---
+# Add a network static route
```
Examples now do this:
```yaml
- name: Add a static network route
  win_route:
    destination_ip: 192.168.2.10
    ...
```

dagwieers commented on lib/ansible/modules/windows/win_route.py:
```diff
@@ -0,0 +1,84 @@
+    description:
+      - Destination IP address
+    required: true
+    default: none
```
If a parameter is required, adding the `default:` option is not needed and undesired.

dagwieers commented on lib/ansible/modules/windows/win_route.py:
```diff
@@ -0,0 +1,84 @@
+  metric:
+    description:
+        - Metric used by the static route
+    required: false
```
Options that are not required, don't need `required: false`, it's implicit.

dagwieers commented on lib/ansible/modules/windows/win_route.py:
```diff
@@ -0,0 +1,84 @@
+  state:
+    description:
+      - present or absent
+    required: false
```
Same here, ditch the `required: false` !

dagwieers commented on lib/ansible/modules/windows/win_route.py:
```diff
@@ -0,0 +1,84 @@
+    default: none
+  subnet_mask:
+    description:
+        - Subent mask of the destination IP address
```
Typo, should be *Subnet mask*.

dagwieers commented on lib/ansible/modules/windows/win_route.py:
```diff
@@ -0,0 +1,84 @@
+version_added: "2.4"
+short_description: Add or remove a static route.
+description:
+    - Add or remove a static route. Powershell 3.0 is needed
```
Add a trailing dot here.
Or rather, remove the Powershell 3.0 requirement, as that's what we require for Ansible anyway !

dagwieers commented on lib/ansible/modules/windows/win_route.py:
```diff
@@ -0,0 +1,84 @@
+    default: 1
+  state:
+    description:
+      - present or absent
```
That's not a proper description. Look at another module to steal a better description.

Commit:
genaral code review

dagwieers commented on lib/ansible/modules/windows/win_route.ps1:
```diff
@@ -0,0 +1,174 @@
+
+# win_route (Add or remove a network static route)
+
+$params = Parse-Args $args -supports_check_mode $true
```
Your code now indicates that the module supports check-mode, but you are not defining the $check_mode variable, and are not doing anything to support check-mode in your code. So this is worse, because people will expect that in check-mode no real changes are being made, but instead changes will be made !

dagwieers commented on lib/ansible/modules/windows/win_route.ps1:
```diff
@@ -0,0 +1,174 @@
+
+Function Test-SubnetMask {
+    Param (
+        [Parameter(Mandatory=$True)]
```
Why use both $true and $True ? We don't enforce one specific form, but we do expect a module to be consistent in its use.

dagwieers commented on lib/ansible/modules/windows/win_route.py:
```diff
@@ -0,0 +1,79 @@
+  gateway:
+    description:
+        - Gateway used by the static route
+    required: true
```
Why is a gateway required ?
One can add local routes (to local networks).

dlazz replied:
Gateway is required by route ADD command. Gateway is not only intended as 'Internet gateway' but has to be used to route packet to different subnet (private or public).

dagwieers replied:
Right, that's awkward :-)

Local routes require a gateway of 0.0.0.0, so what I would do is the following in this case:
- Make gateway optional
- If it is set to $null, use 0.0.0.0 instead

IMO that's better than having it default to 0.0.0.0 (although that still works too).

PS Local routes show as On-link

Commit:
changed 'supported_by' in doc, supports_check_mode removed

dagwieers commented:
@nitzmahone @jborean93 @jhawkesworth What do you think ?

It's probably better to have a pure-powershell implementation that is idempotent using New-NetRoute, Get-NetRoute, Set-NetRoute, Remote-NetRoute ?

jhawkesworth commented:
Its a good point. In this case, it looks like the Cmdlets ( *-NetRoute ) are capable of doing what the  .exe can, in which case I'd probably say it is preferable to use the Cmdlets and not wrap the .exe.  It appears the Cmdlets were available in powershell 3.0 as well, which is another consideration.

In this case using the Cmdlets would also make adding check mode support easier as you can use the -WhatIf parameter on these Cmdlets.

dlazz commented:
>local routes require a gateway of 0.0.0.0, so what I would do is the following in this case:
> Make gateway optional
> If it is set to $null, use 0.0.0.0 instead

Ok, maybe this can be a better solution, also because gateway is not mandatory when you delete route. :)
 

> It's probably better to have a pure-powershell implementation that is idempotent using New-NetRoute, Get-NetRoute, Set-NetRoute, Remote-NetRoute ? 

I decided to use route command instead of pure-powershell cmdlets to avoid user interaction. When you want to remove a route using **Remote-NetRoute** you're prompted for confirmation and there's no way to override it with a -Force parameter or something like that.
I thought that using route command only for deletion wasn't consistent.
**Get-CimInstance win32_ip4PersistedrouteTable**  was used because return nothing if don't find any route, while **Get-NetRoute** return an error.

dagwieers commented:
@dlazz I understand. Not being able to use Remove-NetRoute non-interactively seems another mistake by Microsoft. Sure people can shut themselves out, but there's more than one way to do that. Silly...

dagwieers commented:
@dlazz I just tested, and if you add `-Confirm:$false` you can run **Remove-NetRoute** non-interactively !

dlazz commented:
@dagwieers, excellent! At this point I've got no excuses to implement it :)
It's a bit more tricky 'cause Add-NetRoute uses network prefix, so I've to convert input subnet mask. 
I prefer to use subnet mask 'cause I think is more pratical than network prefix.

dagwieers commented:
@dlazz It's best to use what other modules are doing as well. But in any case we have jinja2 filters to convert from one format to another.

http://docs.ansible.com/ansible/playbooks_filters_ipaddr.html

And for those conversions we don't have filters yet, it's easy to add them later.

dlazz commented:
@dagwieers ok. Looking at nmcli module, ip address is implemented using the format address/prefix:

http://docs.ansible.com/ansible/nmcli_module.html 

Do you think is a more advisable way?

dagwieers commented:
I was involved with **vmware_guest** and there it was decided to use `ip:` and `netmask:`.
I have no strong preference, but maybe we do want to make it consistent for all Windows modules ?

@dlazz We have a Windows Working Group meeting later today/tomorrow ! https://github.com/ansible/community/blob/master/MEETINGS.md#schedule 

@nitzmahone @jborean93 @jhawkesworth Back to you :-) Do we have a preference here ?

dlazz commented:
> @dlazz We have a Windows Working Group meeting later today/tomorrow ! 
@dagwieers Thanks for your invitation but it's a bit late for me ;) Maybe next time!

dlazz commented:
I'm trying to use New-NetRoute cmdlet.
When I run the new function directly in powershell everything works fine, but when I run it with ansible this error is returned:

`An exception occurred during task execution. To see the full traceback, use -vvv. The error was: +                        ~~~
fatal: [default]: FAILED! => {"changed": false, "failed": true, "msg": "At C:\\Users\\vagrant\\AppData\\Local\\Temp\\ansible-tmp-1491839669.11-51240291159332\\win_route.ps1:368 char:24\r\n+ if ($state -eq \"present\"){\r\n+                        ~~~\nThe string is missing the terminator: \".\r\n\r\nAt C:\\Users\\vagrant\\AppData\\Local\\Temp\\ansible-tmp-1491839669.11-51240291159332\\win_route.ps1:311 char:9\r\n+     try {\r\n+         ~\nMissing closing '}' in statement block.\r\n\r\nAt C:\\Users\\vagrant\\AppData\\Local\\Temp\\ansible-tmp-1491839669.11-51240291159332\\win_route.ps1:380 char:1\r\n+ \nThe Try statement is missing its Catch or Finally block.\r\n\r\nAt C:\\Users\\vagrant\\AppData\\Local\\Temp\\ansible-tmp-1491839669.11-51240291159332\\win_route.ps1:310 char:17\r\n+   if (!($Route)){\r\n+                 ~\nMissing closing '}' in statement block.\r\n\r\nAt C:\\Users\\vagrant\\AppData\\Local\\Temp\\ansible-tmp-1491839669.11-51240291159332\\win_route.ps1:297 char:20\r\n+ Function Add-Route {\r\n+                    ~\nMissing closing '}' in statement block.\r\n\r\n"}`

I checked the code several time, but everything seems to be in the right place.
Maybe I miss something, but I can't get it.
For convenience I attach the code instead of pull it.

[win_route.ps1.zip](https://github.com/ansible/ansible/files/910933/win_route.ps1.zip)

Any suggestion?

dlazz commented:
I forgot to say that now destination_ip is in ip/prefix format as required by powershell cmdlet.

dagwieers commented:
@dlazz We discussed it briefly this morning and the consensus was to use CIDR format for modules going forward. We are going to make this proposal project-wide as well. Also, since the parameter is used for more than single IP addresses (e.g. subnet, list of IPs), it is probably better to name it "destination" instead.

In the future we expect a parameter-type "cidr" to be used, but today it does not exist (not for python, and not for powershell modules).

Let me know if you need any more help.

Commit:
use of powershell cmdlet instead of route command, destnation in CIDR format, adds check mode support

dlazz commented:
I reviewed all the code. 
After hour of going mad, I noticed that a dash before a parameter was different from other dashes. Apparently this one  – and this - are not the same thing!
Now all commands are in pure poweshell, a *destination* parameter is used with CIDR format, destination_ip and subnet_mask was removed and check mode is added.

dagwieers reviewed (approved):
shipit

dagwieers commented:
shipit

Commit:
adds win_psmodule module

Commit:
documentation review

Commit:
documentation review

Commit:
removed files accidentatlly pushed

dlazz commented:
Sorry, I accidentally pushed files of another module I'm working on.

dagwieers commented:
shipit

cc: @nitzmahone @jborean93 @jhawkesworth Your input is appreciated !

dagwieers commented:
Ok, so what is the status now ? Is this ready to be merged ?

Commit:
add integration tests, add return in documentation

Commit:
add conditional statement in test, add os requirementes in documentation

Commit:
fix documentation

Commit:
fix os check in tests

dlazz commented:
I added integration tests and discovered that the module doesn't work with os pre windows 2012 R2  because ```Find-NetRoute``` cmdlet was introduced only in this os version. I added this requirement in documentation too. I wonder if use ```route``` dos command, instead of pure powershell, could be a better way to add operating system retro compatibility.

dagwieers commented:
@dlazz It's perfectly fine to not support older Windows versions (which keeps the module easier to maintain and clean). However if you need support for older Windows versions, or you prefer to keep that around (since you basically already wrote that logic), you can.

Beware though that for support, you probably want to add some notes to the users about the difference for different Windows versions. It's best to ensure that the options and output of the module should be identical for older ander newer Windows versions.

nitzmahone commented:
Only two minor beefs with this, but I'll merge as-is:
- the module should have an explicit check (+ coverage by integration tests) to fail gracefully if the cmdlets are missing (eg, target OS is too old)
- lack of IPv6 support - this could be added later, and I *think* since we're only dealing with addresses and not address+port, we *should* be able to unambiguously tell if an address is IPv6 or not by simple inspection, so we shouldn't need separate args for IPv6 support down the line. I guess we can always alias these to, eg, `ipv4_gateway` if for some reason we do.
